File: options.py
import math;
import sys
import config
    
def roundup(value, precision):
    rval = round(value, precision)
    tip = 1 / math.pow(10, precision)
    if value > rval:
        rval = rval + tip
    return rval
    
def calcOptionTradeCost(contracts, costPerTrade = 6.95, costPerContract = 0.75):
    return costPerTrade + (contracts * costPerContract)
    
def optionPrice(price, contracts = 1):
    return price * 100 * contracts
    
def calcCostPerOptionContract(buyPrice, costPerContract = 0.75):
    return optionPrice(buyPrice) + costPerContract
    
def contributionMarginPerOptionContract(buyPrice, sellPrice, costPerContract = 0.75):
    return optionPrice((sellPrice - buyPrice)) - costPerContract

def calcTotalCost(buyPrice, contracts, costPerTrade = 6.95, costPerContract = 0.75):
    return optionPrice(buyPrice, contracts) + calcOptionTradeCost(contracts, costPerTrade, costPerContract)
    
def calcNetRevenue(sellPrice, contracts, costPerTrade = 6.95, costPerContract = 0.75):
    return optionPrice(sellPrice, contracts) - calcOptionTradeCost(contracts, costPerTrade, costPerContract)
    
def calcNetProfitOnOption(buyPrice, sellPrice, contracts, costPerTrade = 6.95, costPerContract = 0.75):
    totalTradeCost = 2 * calcOptionTradeCost(contracts, costPerTrade, costPerContract)
    grossProfit = optionPrice((sellPrice - buyPrice), contracts)
    return roundup(grossProfit - totalTradeCost, 2)
    
def optionContractVolumeToRealizeProfit(buyPrice, sellPrice, profit = 0.0, costPerTrade = 6.95, costPerContract = 0.75):
    return math.ceil((2 * costPerTrade + profit)/(contributionMarginPerOptionContract(buyPrice, sellPrice, costPerContract) - costPerContract))
    
def calcMinimumNumberOfContractsToBreakEven(buyPrice, sellPrice, costPerTrade = 6.95, costPerContract = 0.75):
    return math.ceil(2 * costPerTrade/(contributionMarginPerOptionContract(buyPrice, sellPrice, costPerContract) - costPerContract))
    
def calcMinimumContractSellPriceToBreakEven(buyPrice, contracts, costPerTrade = 6.95, costPerContract = 0.75):
    totalTradeCost = 2 * calcOptionTradeCost(contracts, costPerTrade, costPerContract)
    return roundup((totalTradeCost + optionPrice(buyPrice, contracts))/(contracts * 100), 2)
    
def calcMinimumOptionInvestmentToRealizeGain(buyPrice, sellPrice, costPerTrade = 6.95, costPerContract = 0.75):
    contracts = calcMinimumNumberOfContractsToBreakEven(buyPrice, sellPrice, costPerTrade, costPerContract)
    totalCost = calcTotalCost(buyPrice, contracts, costPerTrade, costPerContract)
    return (totalCost, contracts)
    
def calcMinimumOptionInvestmentToRealizeProfit(buyPrice, sellPrice, profit, costPerTrade = 6.95, costPerContract = 0.75):
    contracts = optionContractVolumeToRealizeProfit(buyPrice, sellPrice, profit, costPerTrade, costPerContract)
    totalCost = calcTotalCost(buyPrice, contracts, costPerTrade, costPerContract)
    return (totalCost, contracts)
    
def calcNumberOfOptionsPurchasable(buyPrice, cash, costPerTrade = 6.95, costPerContract = 0.75):
    itemCost = calcCostPerOptionContract(buyPrice, costPerContract)
    availableAssets = cash - costPerTrade;
    return math.floor(availableAssets / itemCost)
        
if __name__ == '__main__':
    '''
        call like 
        python options.py [buyPrice] [sellPrice] [contracts]        # Calculates profit from trade info
        python options.py [-p] [buyPrice] [sellPrice] [profit]      # Calculates number of contracts to achieve target profit level from range 
        python options.py [-t] [buyPrice] [contracts] [profit]      # Calculates target sell price to achieve break even and target profit level
        python options.py [-m] [buyPrice] [cashToInvest]            # Calculates the maximum number of contracts you can purchase given the buyPrice and cash        
    '''
    buyPrice = float(sys.argv[2])
    profit = float(sys.argv[4]) if len(sys.argv) > 4 else 0
    
    if sys.argv[1] == "-p":
        sellPrice = float(sys.argv[3])
                         
        minimumInvestment, minimumContracts = calcMinimumOptionInvestmentToRealizeGain(buyPrice, sellPrice)
        print "Buy: %.2f Sell: %.2f Break Even Investment: %.2f Contracts: %d Net Profit %.2f" % (buyPrice, sellPrice, minimumInvestment, minimumContracts, calcNetProfitOnOption(buyPrice, sellPrice, minimumContracts))
        
        minimumInvestment, minimumContracts = calcMinimumOptionInvestmentToRealizeProfit(buyPrice, sellPrice, profit)
        print "Buy: %.2f Sell: %.2f Profit Target Investment: %.2f Contracts: %d Profit: %.2f" % (buyPrice, sellPrice, minimumInvestment, minimumContracts, calcNetProfitOnOption(buyPrice, sellPrice, minimumContracts))
    elif sys.argv[1] == "-t":
        contracts = float(sys.argv[3])

        sellPrice = calcMinimumContractSellPriceToBreakEven(buyPrice, contracts)
        print "Buy: %.2f Break Even Sell: %.2f Contracts: %d Net Profit %.2f" % (buyPrice, sellPrice, contracts, calcNetProfitOnOption(buyPrice, sellPrice, contracts))
        
        minimumInvestment, minimumContracts = calcMinimumOptionInvestmentToRealizeProfit(buyPrice, sellPrice, profit)
        print "Buy: %.2f Sell: %.2f Profit Target Investment: %.2f Contracts: %d Profit: %.2f" % (buyPrice, sellPrice, minimumInvestment, minimumContracts, calcNetProfitOnOption(buyPrice, sellPrice, minimumContracts))
    elif sys.argv[1] == "-m":
        cash = float(sys.argv[3])
        contracts = calcNumberOfOptionsPurchasable(buyPrice, cash)
        totalCost = calcTotalCost(buyPrice, contracts)
        
        print "You can buy %d Contracts at %.2f for %.2f" % (contracts, buyPrice, totalCost)
    else:
        buyPrice = float(sys.argv[1])
        sellPrice = float(sys.argv[2])
        contracts = float(sys.argv[3])
        
        print "Buy: %.2f Sell: %.2f Contracts: %d Net Profit %.2f" % (buyPrice, sellPrice, contracts, calcNetProfitOnOption(buyPrice, sellPrice, contracts))

File: options.scala
def round(value: Double, precision: Int): Double = {
    if(precision < 0) 0
    else {
        val factor = Math.pow(10, precision)
        val rval =  Math.round(value * factor) / factor
        val tip = 1 / factor
        if(rval < value) return rval + tip
        else return rval
    }
}

def optionTradeCost(contracts: Int, costPerTrade: Double = 6.95, costPerContract: Double = 0.75): Double = {
    return costPerTrade + (contracts * costPerContract)
}

def optionTradePrice(price: Double, contracts: Int = 1): Double = {
    return price * 100 * contracts;
}

def costPerOptionContract(buyPrice: Double, costPerContract: Double = 0.75): Double = {
    return optionTradePrice(buyPrice) + costPerContract
}

def contributionMarginPerOptionContract(buyPrice: Double, sellPrice: Double, costPerContract: Double = 0.75): Double = {
    return optionTradePrice((sellPrice - buyPrice)) - costPerContract
}

def totalOptionTradeCost(buyPrice: Double, contracts: Int, costPerTrade: Double = 6.95, costPerContract: Double = 0.75): Double = {
    return optionTradePrice(buyPrice, contracts) + optionTradeCost(contracts, costPerTrade, costPerContract)
}

def netOptionTradeRevenue(sellPrice: Double, contracts: Int, costPerTrade: Double = 6.95, costPerContract: Double = 0.75): Double = {
    return optionTradePrice(sellPrice, contracts) - optionTradeCost(contracts, costPerTrade, costPerContract)
}

def netOptionTradeProfit(buyPrice: Double, sellPrice: Double, contracts: Int, costPerTrade: Double = 6.95, costPerContract: Double = 0.75): Double = {
    val totalTradeCost = 2 * optionTradeCost(contracts, costPerTrade, costPerContract)
    val grossProfit = optionTradePrice((sellPrice - buyPrice), contracts)
    return round(grossProfit - totalTradeCost,2)
}

def breakEvenOptionSellPrice(buyPrice: Double, contracts: Int, costPerTrade: Double = 6.95, costPerContract: Double = 0.75): Double = {
    val totalTradeCost = 2 * optionTradeCost(contracts, costPerTrade, costPerContract)
    return round((totalTradeCost + optionTradePrice(buyPrice, contracts))/(contracts * 100), 2)
}

def optionContractVolumeToRealizeProfit(buyPrice: Double, sellPrice: Double, profit:Double = 0.0, costPerTrade: Double = 6.95, costPerContract: Double = 0.75): Int = {
    return Math.ceil((2 * costPerTrade + profit)/(contributionMarginPerOptionContract(buyPrice, sellPrice, costPerContract) - costPerContract)).toInt
}

def breakEvenOptionContractVolume(buyPrice: Double, sellPrice: Double, costPerTrade: Double = 6.95, costPerContract: Double = 0.75): Int = {
    return Math.ceil(2 * costPerTrade/(contributionMarginPerOptionContract(buyPrice, sellPrice, costPerContract) - costPerContract)).toInt
}

def maxOptionContractsPurchasable(buyPrice: Double, cash: Double, costPerTrade: Double = 6.95, costPerContract: Double = 0.75): Int = {
    val itemCost = costPerOptionContract(buyPrice, costPerContract)
    val availableAssets = cash - costPerTrade;
    return Math.floor(availableAssets / itemCost).toInt
}

def minInvestmentToRealizeGainOnOptionTrade(buyPrice: Double, sellPrice: Double, costPerTrade: Double = 6.95, costPerContract: Double = 0.75): Tuple2[Double, Int] = {
    val contracts = breakEvenOptionContractVolume(buyPrice, sellPrice, costPerTrade, costPerContract)
    val totalCost = totalOptionTradeCost(buyPrice, contracts, costPerTrade, costPerContract)
    return new Tuple2[Double, Int](totalCost, contracts)
}

def minInvestmentToRealizeProfitOnOptionTrade(buyPrice: Double, sellPrice: Double, profit: Double, costPerTrade: Double = 6.95, costPerContract: Double = 0.75): Tuple2[Double, Int] = {
    val contracts = optionContractVolumeToRealizeProfit(buyPrice, sellPrice, profit, costPerTrade, costPerContract)
    val totalCost = totalOptionTradeCost(buyPrice, contracts, costPerTrade, costPerContract)
    return new Tuple2[Double, Int](totalCost, contracts)
}

println(optionTradeCost(1))
println(optionTradePrice(.50))
println(costPerOptionContract(.50))
println(totalOptionTradeCost(.50, 1))
println(netOptionTradeRevenue(.50,1))
println(netOptionTradeProfit(.20, .50, 1))
println(breakEvenOptionSellPrice(.20, 1))
println(netOptionTradeProfit(.20, .36, 1))
println(breakEvenOptionContractVolume(.20, .36))
println(breakEvenOptionContractVolume(.20, .35))
println(netOptionTradeProfit(.20, .35, 2))
println(maxOptionContractsPurchasable(.20, 100))
println(minInvestmentToRealizeGainOnOptionTrade(.20, .35))
println(minInvestmentToRealizeProfitOnOptionTrade(.20, .35, 20.0))
println(netOptionTradeProfit(.20, .35, 3))
